src/Microsoft.AspNet.Server.Kestrel/Infrastructure/ThreadPoolActions.cs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Server.Kestrel.Http;
+
+namespace Microsoft.AspNet.Server.Kestrel.Infrastructure
+{
+    public class ThreadPoolActions
+    {
+        private readonly IKestrelTrace _log;
+
+        private readonly WaitCallback _runAction;
+        private readonly WaitCallback _completeTcs;
+        private readonly WaitCallback _abortFrame;
+        private readonly WaitCallback _returnBlocks;
+        private readonly WaitCallback _returnWrittenBlocks;
+
+        public ThreadPoolActions(IKestrelTrace log)
+        {
+            _log = log;
+
+            // Curry and capture log in closures once
+            _runAction = (o) =>
+            {
+                try
+                {
+                    ((Action)o)();
+                }
+                catch (Exception e)
+                {
+                    _log.ApplicationError(e);
+                }
+            };
+
+            _completeTcs = (o) =>
+            {
+                try
+                {
+                    ((TaskCompletionSource<object>)o).SetResult(null);
+                }
+                catch (Exception e)
+                {
+                    _log.ApplicationError(e);
+                }
+            };
+
+            _abortFrame = (o) =>
+            {
+                try
+                {
+                    ((Frame)o).Abort();
+                }
+                catch (Exception e)
+                {
+                    _log.ApplicationError(e);
+                }
+            };
+
+            _returnBlocks = (o) =>
+            {
+                try
+                {
+                    var block = ((MemoryPoolBlock2)o);
+                    ReturnBlocksInternal(block);
+                }
+                catch (Exception e)
+                {
+                    _log.ApplicationError(e);
+                }
+            };
+
+            _returnWrittenBlocks = (o) =>
+            {
+                try
+                {
+                    var block = ((MemoryPoolBlock2)o);
+                    ReturnWrittenBlocksInternal(block);
+                }
+                catch (Exception e)
+                {
+                    _log.ApplicationError(e);
+                }
+            };
+        }
+
+        private static void ReturnBlocksInternal(MemoryPoolBlock2 block)
+        {
+            while (block != null)
+            {
+                var returningBlock = block;
+                block = returningBlock.Next;
+
+                returningBlock.Pool?.Return(returningBlock);
+            }
+        }
+
+        private static void ReturnWrittenBlocksInternal(MemoryPoolBlock2 block)
```

Comment:
I don't like moving the WaitCallback implementations outside of the class queuing the work. I noticed we were doing this with Frame.Abort in the previous PR and probably should have brought it up then.

I know this is being done to avoid per-call closure allocations, but I think think can be done by making the state parameter include the logger we could achieve the same thing.